File: snippets/kotlin_mpp_lib/shared/src/commonMain/kotlin/com/example/kotlinmpplib/SendOnchain.kt
package com.example.kotlinmpplib

import breez_sdk.*
class SendOnchain {
    fun getCurrentRevSwapLimits(sdk: BlockingBreezServices) {
        // ANCHOR: get-current-reverse-swap-limits
        try {
            val currentLimits = sdk.onchainPaymentLimits()
            // Log.v("Breez", "Minimum amount, in sats: ${currentLimits.minSat}")
            // Log.v("Breez", "Maximum amount, in sats: ${currentLimits.maxSat}")
        } catch (e: Exception) {
            // handle error
        }
        // ANCHOR_END: get-current-reverse-swap-limits
    }

    fun preparePayOnchain(sdk: BlockingBreezServices, currentLimits: OnchainPaymentLimitsResponse) {
        // ANCHOR: prepare-pay-onchain
        val amountSat = currentLimits.minSat
        val satPerVbyte = 10.toUInt()
        try {
            val prepareRequest = PrepareOnchainPaymentRequest(amountSat, SwapAmountType.SEND, satPerVbyte)
            val prepareRes = sdk.prepareOnchainPayment(prepareRequest)
            // Log.v("Breez", "Sender amount: ${prepareRes.senderAmountSat} sats")
            // Log.v("Breez", "Recipient amount: ${prepareRes.recipientAmountSat} sats")
            // Log.v("Breez", "Total fees: ${prepareRes.totalFees} sats")
        } catch (e: Exception) {
            // handle error
        }
        // ANCHOR_END: prepare-pay-onchain
    }

    fun startReverseSwap(sdk: BlockingBreezServices, prepareRes: PrepareOnchainPaymentResponse) {
        // ANCHOR: start-reverse-swap
        val address = "bc1.."
        try {
            sdk.payOnchain(PayOnchainRequest(address, prepareRes))
        } catch (e: Exception) {
            // handle error
        }
        // ANCHOR_END: start-reverse-swap
    }

    fun checkRevSwapStatus(sdk: BlockingBreezServices) {
        // ANCHOR: check-reverse-swaps-status
        for (rs in sdk.inProgressOnchainPayments()) {
            // Log.v("Breez", "Onchain payment ${rs.id} in progress, status is ${rs.status}")
        }
        // ANCHOR_END: check-reverse-swaps-status
    }
}

File: snippets/swift/BreezSDKExamples/Sources/SendOnchain.swift
//
//  SendOnchain.swift
//
//
//  Created by ruben on 14/11/2023.
//

import BreezSDK
import Foundation

func GetCurrentLimits(sdk: BlockingBreezServices) -> OnchainPaymentLimitsResponse?? {
    //  ANCHOR: get-current-reverse-swap-limits
    let currentLimits = try? sdk.onchainPaymentLimits()
    if let limits = currentLimits {
        print("Minimum amount, in sats: \(limits.minSat)")
        print("Maximum amount, in sats: \(limits.maxSat)")
    }
    // ANCHOR_END: get-current-reverse-swap-limits
    return currentLimits
}

func PreparePayOnchain(sdk: BlockingBreezServices, currentLimits: OnchainPaymentLimitsResponse) -> PrepareOnchainPaymentResponse? {
    // ANCHOR: prepare-pay-onchain
    let amountSat = currentLimits.minSat
    let satPerVbyte: UInt32 = 5

    let prepareRequest = PrepareOnchainPaymentRequest(amountSat: amountSat, amountType: .send, claimTxFeerate: satPerVbyte);
    let prepareResponse = try? sdk.prepareOnchainPayment(req: prepareRequest)

    if let response = prepareResponse {
        print("Sender amount, in sats: \(response.senderAmountSat)")
        print("Recipient amount, in sats: \(response.recipientAmountSat)")
        print("Total fees, in sats: \(response.totalFees)")
    }
    // ANCHOR_END: prepare-pay-onchain
    return prepareResponse
}

func StartReverseSwap(sdk: BlockingBreezServices, prepareResponse: PrepareOnchainPaymentResponse) -> PayOnchainResponse? {
    // ANCHOR: start-reverse-swap
    let destinationAddress = "bc1.."

    let response = try? sdk.payOnchain(req: PayOnchainRequest(recipientAddress: destinationAddress, prepareRes: prepareResponse))
    // ANCHOR_END: start-reverse-swap
    return response
}

func checkReverseSwap(sdk: BlockingBreezServices) {
    // ANCHOR: check-reverse-swaps-status
    if let inProgressOnchainPayments = try? sdk.inProgressOnchainPayments() {
        for rs in inProgressOnchainPayments {
            print("Onchain payment \(rs.id) in progress, status is \(rs.status)")
        }
    }
    // ANCHOR_END: check-reverse-swaps-status
}
